### DISCUSSION TITLE
Technical documentation - Chrome extension

### DISCUSSION THREAD

**Discussion by AdityaPimpalkar** [twentyhq/twenty#5264] [Engineering vision]
## Core principles

- DOM interaction
- Secure communication
- Clear workflow for user interaction and data flow
- Robust Testing Strategy
- Security Best Practices

## Definitions

**Content scripts**: Scripts that read and modify the content of a page, can make changes to their JavaScript environment without conflicting with their host page or other extensions content scripts. It can use the standard HTML DOM to read and change the content of a page.

**Service workers:** Scripts which can monitor browser events in the background, these scripts are special JavaScript environments that only handle chrome events and terminate when they're not needed - like when we install extensions for the first time, when we switch tabs, or handle message events. These do have access to the DOM object and so cannot manipulate it. 

**Event messaging API**: For communication, we can use the chromes built in messaging API which allows both service workers and content scripts to listen for each other's messages and respond on the same channel. There are two types here - one for one-time requests, and a more complex one for long-lived connections that allow multiple messages to be sent. We can have the message listeners on both the extensions service workers and content scripts.

<img width="994" alt="image4" src="https://github.com/twentyhq/twenty/assets/13139771/bbba0d7d-7a41-42c4-b6f7-9f53133f97b4">

### Scenario 1 - 
Let's consider that a user is viewing someone's linkedin profile and wants to save their details to Twenty. 

1. We can inject a button to the DOM using content scripts, which lets you save those persons details.
2. If the user chooses to press the button, we send an action message event to the service worker passing the relevant information as input.
3. The service worker will listen and receive said event, sanitize the input and then make a database call, after that we can do a callback as the response of the event performed.
4. Display the user that information was saved.

<img width="952" alt="image3" src="https://github.com/twentyhq/twenty/assets/13139771/ec1ef76b-92da-4b8f-a5ae-122b8989074c">

### Scenario 2 -
Let's consider we are using the React UI to save someone's linkedin profile and want to save their details to Twenty.

1. To render text from current pages DOM into react, we can make a request with a message event to a content script
2. Since the content script has access to DOM and can also have listeners present, we can pass relevant information as JSON response in the callback to React. 
3. React will sanitize the input and then make required database calls

<img width="742" alt="image2" src="https://github.com/twentyhq/twenty/assets/13139771/24bf3a09-2a4a-4b7b-a0bd-dc9c3e846d30">

## Testing

Possible to use puppeteer or playwright for end-to-end testing 
https://developer.chrome.com/docs/extensions/how-to/test/end-to-end-testing

## Security

The documentation mentions about content scripts being not immune to attacks - https://developer.chrome.com/docs/extensions/develop/security-privacy/stay-secure#content_scripts

Also about sanitizing and validating inputs - 
https://developer.chrome.com/docs/extensions/develop/security-privacy/stay-secure#sanitize

Note: we can make backend/database calls from content scripts too, but since they can be prone to malicious attacks due to having access to DOM, it's considered a good practice to have them written on service workers.

**Comment by neilk17** [comment]
Hey Aditya, 
I've self deployed Twenty and I want to use the chrome extension but it doesn't let me go beyond this page. Once I enter my URL it doesn't save it.

<img width="479" alt="Screenshot 2024-06-28 at 11 09 22 AM" src="https://github.com/twentyhq/twenty/assets/25948274/5162e272-ed29-491c-a42f-050f9f8791ce">

**Comment by JhumanJ** [comment]
Hey, I created a small Linkedin extension for Twenty (compatible with self hosted instances). It's not on the chrome store but you can check it out and try it here if you're interested: https://github.com/JhumanJ/twenty-crm-extension. It's great for 1 click Linkedin capture.
<img width="1008" height="642" alt="image" src="https://github.com/user-attachments/assets/8f6cbd48-4abc-48b9-8b1d-225d537c2dd9" />
